### ISSUE TITLE
VMF loss problems

### ISSUE THREAD

**Issue by Islanna** [Sachin19/seq2seq-con#2]
Hi,

I was trying to reproduce the Von Mises-Fisher loss  from the paper following your code and got some unexpected results after experiments with fasttext embeddings(m=300). Most of the problems comes from the first loss term - `logcmk`.

1. Exact loss value is equal to ~ -400. I suppose, there is no requirement for the loss values to be non-negative, since we are searching for the minimum, but it seems pretty weird. Also, approximated loss value is ~ 290.
2. Regardless of the similarity between output and target vectors gradients after loss.backward() are in the narrow range. For example, if we calculate the loss between output=target and target, the gradient values are ~2e-05, while the gradient between completely different output and target gives the gradient values ~6e-05. The usual cosine loss for the same example gives ~e-21 and ~5e-05 respectively.

I've tried to play with the regularization and some other parameters, but haven't got rif of the problems above.  Is it a correct behavior for the vmfloss? What could be a problem? I'd really appreciate your help.

Also, I haven't figured out from the paper: is there a reason for using `||e_hat||` as `kappa`?

Thanks a lot!
Anna

**Comment by Sachin19** [comment]
Hi,

1. The loss value is negative because it is the log of probability density as opposed to a probability distribution in case of softmax. In any case, it has no bearing on the model performance. Approximated loss seems a little weird, in all my experiments in remained in the range of -400. Could you share more details of your experiment?

2. That is infact true for later epochs when the model is almost converging, but I didn't observe this behavior in the early epochs. Again, more details about your experiments would help in resolving this. 

`||e_hat||` was taken as `kappa` to incorporate `kappa` with the output vector itself without adding extra parameters and it worked well experiments. In hindsight, it works as a regularization for cosine loss.
